Initial drafts for Prompt Builder doc (#8111)
microsoft/vscode-docs: docs/intelligentapps/promptbuilder.md

---
Order: 4
Area: intelligentapps
TOCTitle: Prompt Builder
ContentId:
PageTitle: Prompt Builder
DateApproved:
MetaDescription: Get Started with creating, iterating and optimizing your prompts in AI Toolkit.
---
# Prompt Engineering in AI Toolkit

Prompt Builder in AI Toolkit streamlines the prompt engineering workflow by generating starter prompts, helping you iterate and refine with each run, breaking down complex tasks through prompt chaining and structured outputs, and providing easy access to code for seamless LLM integration via APIs.

> ![PromptBuilder demo](./images/promptbuilder/promptbuilder.gif)

Author: Updated

## Create, edit and test prompts

To access the Prompt Builder:
- In AI Toolkit view, select **Prompt Builder**
- Select **Try in Prompt Builder** from a model card in the model catalog

To test a prompt in the prompt builder, follow these steps:
1. In **Models**, select a model from the dropdown list or click **Browser models** to add another model from model catalog.
    ![select_model](./images/promptbuilder/s1_models.png)
2. Enter **System prompt** and **User prompt**
   > [!TIP]
   > - System prompt is optional, it is used to provide instructions with relevant context to guide the model response

>    > - User prompt is mandatory, it is the input you want to send to the model

Reviewer: ```suggestion
```

>    > - If you don't know how to input these prompts, you can simply describe your project idea in natural language and let the AI-powered feature generate prompts for you to experiment with

Reviewer: ```suggestion
   > If you don't know how to input these prompts, you can describe your project idea in natural language and let the AI-powered feature generate prompts for you to experiment with.
```

>    > ![generate_prompts](./images/promptbuilder/generate_prompt.gif)

Reviewer: Add a descriptive alt text

> 3. Click **Run** to send the prompts to the selected model

Reviewer: ```suggestion
3. Select **Run** to send the prompts to the selected model
```

> 4. Optionally, you can click **Add Prompts** to add more User and Assistant prompts to the conversation, or **Use Response as Assistant Prompt** as the history and context you send to the model to further guide the model's behavior

Reviewer: ```suggestion
4. Optionally, select **Add Prompts** to add more user and assistant prompts to the conversation, or select **Use Response as Assistant Prompt** as the history and context you send to the model to further guide the model's behavior
```

> 5. You can repeat the above steps to iterate over your prompts by observing the model response and making changes to the prompts

Reviewer: ```suggestion
5. Repeat the previous steps to iterate over your prompts by observing the model response and making changes to the prompts
```

## Structured output

Structured output support helps you design prompts to deliver outputs in a structured, predictable format.

> ![structured_output](./images/promptbuilder/structured_output.gif)

Reviewer: Alt text

> To test a prompt in the prompt builder, follow these steps:

Reviewer: Is prompt builder a brand or just a feature description? If not a brand, use lowercase everywhere.

Author: It's just a feature

> 1. Click on **Format** dropdown In **Response** area, and select **json_schema**

Reviewer: ```suggestion
1. Select the **Format** dropdown in the **Response** area, and select **json_schema**
```

> 2. Click on **Prepare schema**

Reviewer: ```suggestion
2. Select **Prepare schema**
```

> 3. From the command palette, you have two options, you can either use your own schema by selecting **Select local file**, or use a predefined schema by selecting **Use an example**

Reviewer: ```suggestion
3. From the Command Palette, either select **Select local file** to use your own schema, or select **Use an example** to use a predefined schema
```

> 4. If you proceed with an example, you can select a schema from the dropdown list

Reviewer: ```suggestion
   If you proceed with an example, you can select a schema from the dropdown list
```

> 5. Click on **Run** to send the prompts to the selected model

Reviewer: ```suggestion
5. Select **Run** to send the prompts to the selected model
```

> 6. You can also edit the schema by clicking on **Edit**

Reviewer: ```suggestion
6. You can also edit the schema by selecting **Edit**
```

>    ![edit_schema](./images/promptbuilder/edit_schema.png)

Reviewer: Alt text

## Integrate prompt engineering into your application

> After experimenting with models and prompts, you can get into coding right away by viewing ready-to-use Python code automatically generated

Reviewer: ```suggestion

After experimenting with models and prompts, you can get into coding right away with the automatically generated Python code.
```

> ![view_code](./images/promptbuilder/view_code.gif)

Reviewer: Alt text

> To access the code, follow these steps:

Reviewer: ```suggestion
To view the Python code, follow these steps:
```

> 1. Click on **View Code**

Reviewer: ```suggestion
1. Select **View Code**
```

2. Select the inference SDK you want to use if it's hosted by GitHub

>    > [!TIP]

Reviewer: ```suggestion

```

>    > AI Toolkit will generate the code for the corresponding model you selected using the provider's client SDK. For models hosted by GitHub, you have the option to select the inference SDK you want to use: [`Azure AI Inference SDK`](https://learn.microsoft.com/python/api/overview/azure/ai-inference-readme?view=azure-python-preview) or the SDK from model provider such as [OpenAI SDK](https://platform.openai.com/docs/libraries) or [Mistral API](https://docs.mistral.ai/api).

Reviewer: ```suggestion
   AI Toolkit generates the code for the model you selected by using the provider's client SDK. For models hosted by GitHub, you have the option to select the inference SDK you want to use: [Azure AI Inference SDK](https://learn.microsoft.com/python/api/overview/azure/ai-inference-readme?view=azure-python-preview) or the SDK from the model provider, such as [OpenAI SDK](https://platform.openai.com/docs/libraries) or [Mistral API](https://docs.mistral.ai/api).
```

> 3. You can view the generated code snippet in a new file window and copy them into your application.

Reviewer: ```suggestion
3. The generated code snippet is shown in a new editor, where you can copy it into your application.
```

   > [!TIP]

>    > To authenticate with the model you will normally need API key from the provider. To access models hosted by GitHub, [generate a personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) (PAT) in your GitHub settings

Reviewer: ```suggestion
   > To authenticate with the model you usually need an API key from the provider. To access models hosted by GitHub, [generate a personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) (PAT) in your GitHub settings
```

## Next steps

- [Run a set of prompts](/docs/intelligentapps/bulkrun.md) in an imported dataset, individually or in a full batch

> - [Run evaluation](/docs/intelligentapps/evaluation.md) job for the popular evaluators

Reviewer: ```suggestion
- [Run an evaluation job](/docs/intelligentapps/evaluation.md) for the popular evaluators
```
